src/java/org/apache/cassandra/index/sai/utils/OrderingFilterRangeIterator.java

```diff
@@ -19,76 +19,54 @@
 package org.apache.cassandra.index.sai.utils;
 
 import java.util.ArrayList;
+import java.util.Iterator;
 import java.util.List;
+import java.util.Set;
 import java.util.function.Function;
 import javax.annotation.concurrent.NotThreadSafe;
 
+import org.apache.cassandra.index.sai.SSTableIndex;
 import org.apache.cassandra.io.util.FileUtils;
+import org.apache.cassandra.utils.Pair;
 
 /**
  * An iterator that consumes a chunk of {@link PrimaryKey}s from the {@link RangeIterator}, passes them to the
  * {@link Function} to filter the chunk of {@link PrimaryKey}s and then pass the results to next consumer.
  * The PKs are currently returned in {@link PrimaryKey} order, but that contract may change.
  */
 @NotThreadSafe
-public class OrderingFilterRangeIterator extends RangeIterator
+public class OrderingFilterRangeIterator implements Iterator<Pair<List<OrderIterator>, Set<SSTableIndex>>>, AutoCloseable
```

Comment:
it looks like this is only used as a helper in QC.getTopKRows, can we fold the logic in there instead of going through the structure of an awkward container class?

Reply:
Perhaps we can get rid of this class entirely. It was introduced to prevent us from materializing too many PrimaryKeys at one time for hybrid queries. It is only used for filter then sort query execution. However, if we materialize more than 100k Primary Keys for a query, it seems like we might want to change the execution path and do sort then filter. We could even use the number of materialized PrimaryKeys as feedback into the query planner.

Reply:
Tag it with a VSTODO then and make it Piotr's problem? :)

In which case I suppose you can ignore my complaint about the anonymous Pair.